dap-launch.el

```diff
@@ -0,0 +1,241 @@
+    ;; ("execPath")
+    ;; ("defaultBuildTask")
+    )
+  "Alist of (REGEX . VALUE) pairs listing variables usable in launch.json files.
+This list is iterated from the top to bottom when expanding
+variables in the strings of the selected launch configuration
+from launch.json or in `dap-expand-variable'.
+
+When a REGEX matches (`string-match'), its corresponding VALUE is
+evaluated as follows: if it is a function (or a quoted lambda),
+that function is called with `funcall', and its result, which
+must be a string, is used in place of the variable. If you used
+capture groups in REGEX, the function you specified in VALUE is
+called with the variable as its only argument. This way, you can
+use `string-match' to get the capture groups. If, however, REGEX
+does not contain capture groups, your function is called without
+any arguments. Otherwise, if it is a symbol, the symbol's value
+is used the same way. Lastly, if it is a string, the string is
+used as a replacement. If no regex matches, the empty string is
+used as a replacement and a warning is issued.
+
+See `dap--launch-json-getenv' for an example on how to use
+capture groups in REGEX.")
+
+(defun dap--launch-json-eval-poly-type (value var)
+  "Get the value from VALUE depending on its type.
+If it is a function, and VAR is not nil, call VALUE and pass VAR as an argument.
+If it is a symbol, return its value.
+Otherwise, return VALUE"
+  (cond ((and var (functionp value)) (funcall value var))
+        ((functionp value) (funcall value))
+        ((symbolp value) (symbol-value value))
+        (t value)))
+
+(defun dap--warn-var-nil (var)
+  "Warn that VAR is an unknown launch.json variable."
+  (dap--launch-json-warn-nil
+   (format "variable '%s' is unknown and was ignored" var)))
+
+(defun dap-expand-variable (var)
+  "Expand variable VAR using `dap--launch-json-variables'."
+  (catch 'ret
+    (save-match-data
+      (dolist (var-pair dap-launch-json-variables)
+        (when (string-match (car var-pair) var)
+          (throw 'ret
+                 (or
+                  (dap--launch-json-eval-poly-type
+                   (cdr var-pair)
+                   (if (= (length (match-data)) 2) ;; no capture groups
+                       nil
+                     var
+                     (message "wow capture groups")))
+                  (dap--warn-var-nil var)
+                  "")))))
+    nil))
+
+(defun dap-expand-variables-in-string (s)
+  "Expand all launch.json variables of the from ${variable} in S.
+Return the result."
+  (let ((old-buffer (current-buffer)))
+    (with-temp-buffer
+      (insert s)
+      (goto-char (point-min))
+
+      (save-match-data
+        (while (re-search-forward "${\\([^}]*\\)}" nil t)
+          (let ((var (match-string 1)))
+            (replace-match
+             (with-current-buffer old-buffer
+               (dap-expand-variable var))))))
+
+      (buffer-string))))
+
+(defun dap--launch-json-expand-vars (conf)
```

Comment:
Actually, we can! I will rename the function to dap-launch-configuration-expand-vars (no longer private, perhaps users may want this?) and we can just call (setq launch-args (dap-launch-configuration-expand-vars launch-configuration)) at the start of dap-start-debugging. That would mean that regular templates could use variables (e.g.
```elisp
(dap-register-template ... :executable "${workspaceFolder}/build/gtk-print")
```).

However, merging it into dap-mode.el might not be a good idea. I think having it as a separate file (e.g. dap-variables.el) would be better, because all the variable-code would be in one place and it would make it easier to add new variables. What do you think? Also, how could we load it from dap-mode.el?

Reply:
> However, merging it into dap-mode.el might not be a good idea. I think having it as a separate file (e.g. dap-variables.el) would be better, because all the variable-code would be in one place and it would make it easier to add new variables. What do you think? Also, how could we load it from dap-mode.el?

If you want to structure that as a utility - then I guess it is fine to reverse the dependency. There is also dap-util.el which hosts vscode related stuff atm.

Reply:
That was not the idea, as variable expansion should be integrated into normal configurations. Rather, dap-launch.el would be loaded from dap-mode.el. Or should it better be in dap-mode.el?